### DISCUSSION TITLE
How to run tests with a debugger

### DISCUSSION THREAD

**Discussion by fwh-dc** [openssl/openssl#23073] [Q&A] [answered]
How do I run any of the tests that are executed by "make test" with a debugger?

**Comment by mattcaswell** [comment, marked as answer]
It depends on the test.

Typically check the recipe file (in test/recipes) to see how the underlying test executable is executed and what arguments it expects. Run it using the "shlib_wrap.sh" script. For "sslapi" the recipe looks like this:

````
ok(run(test(["sslapitest", srctop_dir("test", "certs"),
             srctop_file("test", "recipes", "90-test_sslapi_data",
                         "passwd.txt"), $tmpfilename, "default",
             srctop_file("test", "default.cnf"),
             srctop_file("test",
                         "recipes",
                         "90-test_sslapi_data",
                         "dhparams.pem")])),
             "running sslapitest");
````

So run gdb like this (while sitting in the test directory):

````
../util/shlib_wrap.sh gdb --args sslapitest certs recipes/90-test_sslapi_data/passwd.txt tmpfile default default.cnf recipes/90-test_sslapi_data/dhparams.pem
````
You can run a specific subtest using the "-test" argument:

````
../util/shlib_wrap.sh gdb --args sslapitest -test 4 certs recipes/90-test_sslapi_data/passwd.txt tmpfile default default.cnf recipes/90-test_sslapi_data/dhparams.pem
````

And a specific iteration of that test using the "-iter" argument:

````
../util/shlib_wrap.sh gdb --args sslapitest -test 4 -iter 2 certs recipes/90-test_sslapi_data/passwd.txt tmpfile default default.cnf recipes/90-test_sslapi_data/dhparams.pem
````

Some tests are a bit more complicated., e.g.  ssl_test.c (aka ssl_test_new) - there is a separate README for that one (see README.ssltest.md). Running the TLSProxy based tests in gdb is quite tricky but can be done....ask me if you need help with those.

**Reply by fwh-dc** [reply]
Ok thanks. That definitely pushed me in the right direction. I was afraid that the perl scripts did more "stuff".

I did run a couple of different test suites using an IDE. Seems that leaving out `util/shlib_wrap.sh` does not hinder running the tests? Maybe I was lucky.

**Reply by mattcaswell** [reply]
The script just sets up your environment variables to pick up all the code you just built. If you don't have that setup right you might end up using bits of your system version of OpenSSL rather than your locally built version.

That said I debug via an IDE all the time and explicitly set a list of environment variables. In particular I set up my environment like this. I set the OPENSSL_CONF value to a dummy setting, which is usually ok, but sometimes I set it to something explicit depending on what I am trying to do:

````
            "environment": [
                {
                    "name": "LD_LIBRARY_PATH",
                    "value": "${workspaceFolder}"
                },
                {
                    "name": "OPENSSL_CONF",
                    "value": "${workspaceFolder}/test/x.cnf"
                },
                {
                    "name": "OPENSSL_MODULES",
                    "value": "${workspaceFolder}/providers"
                },
                {
                    "name": "OPENSSL_ENGINES",
                    "value": "${workspaceFolder}/engines"
                },
                {
                    "name": "TEST_CERTS_DIR",
                    "value": "${workspaceFolder}/test/certs"
                },
                {
                    "name": "CTLOG_FILE",
                    "value": "${workspaceFolder}/test/ct/log_list.cnf"
                },
                {
                    "name": "OPENSSL_CONF_INCLUDE",
                    "value": "${workspaceFolder}/providers"
                }
            ],
````

**Reply by mattcaswell** [reply]
> I was afraid that the perl scripts did more "stuff".

Sometimes they do (it depends on the test). In particular when it comes to the TLSProxy tests. But mostly it is straightforward.

**Reply by fwh-dc** [reply]
Great thanks for the config, I'll apply that straightaway.

**Reply by mspncp** [reply]
FWIW: instead of `util/shlib_wrap.sh`, I nowadays use `util/wrap.pl`. This Perl script was introduced a while ago by @levitte as an enhancement. AFAIK, it has two advantages, namely that it sets some additional provider-specific environment variables and that it works on Windows, too. But maybe @levitte can give additional information?

**Reply by mspncp** [reply]
Another side note: the wrapper script can have unexpected side effects on the debugger when the system OpenSSL library and the version you are testing are not compatible. This is, as far as I understand it, due to the fact that gdb uses OpenSSH as transport for remote debugging which in turn uses OpenSSL. 

```
../util/shlib_wrap.sh gdb ...
```
